src/DataProtection/DataProtection/src/SP800_108/ManagedSP800_108_CTR_HMACSHA512.cs

```diff
@@ -55,6 +55,55 @@ public static void DeriveKeys(byte[] kdk, ArraySegment<byte> label, ArraySegment
         }
     }
 
+    public static void DeriveKeys(byte[] kdk, ReadOnlySpan<byte> label, ReadOnlySpan<byte> contextHeader, ReadOnlySpan<byte> contextData, Func<byte[], HashAlgorithm> prfFactory, ArraySegment<byte> output)
+    {
+        // make copies so we can mutate these local vars
+        var outputOffset = output.Offset;
+        var outputCount = output.Count;
+
+        var contextSharedLength = contextHeader.Length + contextData.Length;
+
+        using (var prf = prfFactory(kdk))
+        {
+            // See SP800-108, Sec. 5.1 for the format of the input to the PRF routine.
+            var prfInput = new byte[checked(sizeof(uint) /* [i]_2 */ + label.Length + 1 /* 0x00 */ + contextSharedLength + sizeof(uint) /* [K]_2 */)];
+
+            // Copy [L]_2 to prfInput since it's stable over all iterations
+            uint outputSizeInBits = (uint)checked((int)outputCount * 8);
+            prfInput[prfInput.Length - 4] = (byte)(outputSizeInBits >> 24);
+            prfInput[prfInput.Length - 3] = (byte)(outputSizeInBits >> 16);
+            prfInput[prfInput.Length - 2] = (byte)(outputSizeInBits >> 8);
+            prfInput[prfInput.Length - 1] = (byte)(outputSizeInBits);
+
+            // Copy label and context to prfInput since they're stable over all iterations
+            label.CopyTo(prfInput.AsSpan(sizeof(uint)));
+            contextHeader.CopyTo(prfInput.AsSpan(sizeof(uint) + label.Length + 1));
+            contextData.CopyTo(prfInput.AsSpan(sizeof(uint) + label.Length + 1 + contextHeader.Length));
+
+            var prfOutputSizeInBytes = prf.GetDigestSizeInBytes();
+            for (uint i = 1; outputCount > 0; i++)
+            {
+                // Copy [i]_2 to prfInput since it mutates with each iteration
+                prfInput[0] = (byte)(i >> 24);
+                prfInput[1] = (byte)(i >> 16);
+                prfInput[2] = (byte)(i >> 8);
+                prfInput[3] = (byte)(i);
+
+                // Run the PRF and copy the results to the output buffer
+                var prfOutput = prf.ComputeHash(prfInput);
```

Comment:
Is this `HashAlgorithm` (or sublass) ? If so, note that there is a `ComputeHash(byte[], int, int)` from .NET 9 onwards which would allo `prfInput` to be a pooled+oversized array, constrained i.e.

```
// See SP800-108, Sec. 5.1 for the format of the input to the PRF routine.
var prfLength = checked(sizeof(uint) /* [i]_2 */ + label.Length + 1 /* 0x00 */ + contextSharedLength + sizeof(uint) /* [K]_2 */)
#if NET9_0_OR_GREATER // check exact token
var prfInput = ArrayPool<byte>.Shared.Rent(prfLength);
#else
var prfInput = new byte[prfLength];
#endif

// ...

#if NET9_0_OR_GREATER
var prfOutput = prf.ComputeHash(prfInput, 0, prfLength);
ArrayPool<byte>.Shared.Return(prfInput, clearArray: true); // clearing because security-related
#else
var prfOutput = prf.ComputeHash(prfInput);
#endif
```

Reply:
ooh, there's also `TryComputeHash` that allows a buffer to be passed in

Reply:
that's a great suggestion, and I tried it, but it simply throws. There are a couple of allocations that should probably be rented via the ArrayPool, and it would bring the allocations down, but if I replace the code to rent and return, then it fails.

I am going to write down the findings and the problems here in a while

Reply:
Yep; that's why when I use `ArrayPool<T>.Shared.Rent` I almost always call the local `oversized` - then I can't forget! For example: https://github.com/dotnet/extensions/blob/c08790cf79d2d8feaa53751f039f38638741a154/src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/InbuiltTypeSerializer.cs#L43

There's also cheeky things you can do such as:

``` c#
byte[]? lease = null;
Span<byte> buffer = length < 128
    ? stackalloc byte[length]
    : (lease = ArrayPool<byte>.Shared.Rent(length)).AsSpan(0, length);

// code here using "buffer", which is right-sized and could be stack or heap

if (lease is not null)
{
    ArrayPool<byte>.Shared.Return(lease);
}
```

Reply:
thanks for pointing it out, made an optimization